File: LeetCode/1255. Maximum Score Words Formed by Letters/1255.java
class Solution {
    // Helper function to calculate the score of a word
    private int wordScore(String word, int[] score) {
        int wordScore = 0;
        for (char c : word.toCharArray()) {
            wordScore += score[c - 'a'];
        }
        return wordScore;
    }

    // Backtracking function to find the maximum score
    private int backtrack(int index, int currentScore, int[] remainingLetters, String[] words, int[] score) {
        if (index == words.length) {
            return currentScore;
        }

        // Option 1: Skip the current word
        int maxScore = backtrack(index + 1, currentScore, remainingLetters, words, score);

        // Option 2: Include the current word if possible
        String word = words[index];
        int[] wordCount = new int[26];
        for (char c : word.toCharArray()) {
            wordCount[c - 'a']++;
        }

        boolean canUseWord = true;
        for (int i = 0; i < 26; i++) {
            if (wordCount[i] > remainingLetters[i]) {
                canUseWord = false;
                break;
            }
        }

        if (canUseWord) {
            for (int i = 0; i < 26; i++) {
                remainingLetters[i] -= wordCount[i];
            }
            maxScore = Math.max(maxScore, backtrack(index + 1, currentScore + wordScore(word, score), remainingLetters, words, score));
            for (int i = 0; i < 26; i++) {
                remainingLetters[i] += wordCount[i];
            }
        }

        return maxScore;
    }

    public int maxScoreWords(String[] words, char[] letters, int[] score) {
        // Initial call to the backtracking function
        int[] letterCount = new int[26];
        for (char c : letters) {
            letterCount[c - 'a']++;
        }
        return backtrack(0, 0, letterCount, words, score);
    }
}

File: LeetCode/1255. Maximum Score Words Formed by Letters/1255.py
from collections import Counter

class Solution(object):
    def maxScoreWords(self, words, letters, score):
        """
        :type words: List[str]
        :type letters: List[str]
        :type score: List[int]
        :rtype: int
        """
        # Helper function to calculate the score of a word
        def word_score(word):
            return sum(score[ord(char) - ord('a')] for char in word)
        
        # Backtracking function to find the maximum score
        def backtrack(index, current_score, remaining_letters):
            if index == len(words):
                return current_score
            
            # Option 1: Skip the current word
            max_score = backtrack(index + 1, current_score, remaining_letters)
            
            # Option 2: Include the current word if possible
            word = words[index]
            word_count = Counter(word)
            if all(remaining_letters[char] >= word_count[char] for char in word_count):
                new_remaining_letters = remaining_letters - word_count
                max_score = max(max_score, backtrack(index + 1, current_score + word_score(word), new_remaining_letters))
            
            return max_score
        
        # Initial call to the backtracking function
        letter_count = Counter(letters)
        return backtrack(0, 0, letter_count)


